src/data/pfunctor/indexed/basic.lean

```diff
@@ -0,0 +1,319 @@
+/-
+Copyright (c) 2019 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+-/
+import tactic.interactive
+import tactic.mk_opaque
+import control.family
+import control.functor.indexed
+import data.sigma
+
+/-!
+# Polynomial functors between indexed type families
+
+Indexed polynomial functors are used for defining M-types and W-types.
+They map a type family `α : fam J` to the type family
+`λ j, Σ a : A j, B j a ⟶ α`, with `A : fam J` and `B : Π j, A j → fam I`.
+They interact well with Lean's inductive definitions because they
+guarantee that occurrences of `α` are positive.
+-/
+
+universes v v' u u'
+
+open_locale fam
+
+/-- Polynomial functors between indexed type families -/
+structure ipfunctor (I J : Type u) :=
+(A : fam J) (B : Π i, A i → fam I)
```

Comment:
It would be much more natural to use `j` instead of `i` for an index ranging over `J` rather than `I`!